### ISSUE TITLE
Fault tolerant steps may ignore chunks silently

### ISSUE THREAD

**Issue by omarkad2** [spring-projects/spring-batch#3828]
**Bug description**
During the execution of a fault tolerant Step, when a chunk executes successfully, it is tagged as “Completed”, and then, we try to commit the transaction.

In the case where the transaction commit fails with an exception that is not considered as Non-retryable or Fatal (by FaultTolerantStepBuider) like a RuntimeException, the exception handler logs the error in debug (SimpleRetryExceptionHandler#handleException), and we do not put back the chunk in the queue for retry (since tagged Completed) and consequently we skip the chunk silently while having it tagged as “Completed”.

We end up in this case with a "Completed" job with "Completed" chunks and no error stacks.

**Environment**
Spring Batch version: 3.0.10.RELEASE
Java version: 11
database: postgresql
ORM: hibernate

**Steps to reproduce**
- Declare a job with a fault tolerant chunk-oriented step.
- In the item writer make the transaction commit fail with a RuntimeException (for example) by overriding the beforeCommit method with the following:
```
TransactionSynchronizationManager.registerSynchronization(new TransactionSynchronization() {
            @Override
            public void beforeCommit(boolean readOnly) {
                throw new RuntimeException("test");
            }
});
```

**Expected behavior**
Retry the chunk, and if the error persist throw an error and mark the job as failed.

**Comment by benas** [comment]
Thank you for raising this issue. I believe this is a duplicate of #1189. Do you confirm that?

**Comment by omarkad2** [comment]
Hello @benas I can confirm that it is the same issue. Is it something that you consider changing or will it be kept as is ?

**Comment by benas** [comment]
Thank you for your confirmation. I will close this as a duplicate of #1189.

> Is it something that you consider changing or will it be kept as is ?

Both this issue and #1189 are reported against Spring Batch v3, which is not maintained anymore. If this is validated as a bug with a minimal complete example against the current v4.3.1, we will take a look.

**Comment by fmbenhassine** [comment]
@patwix This was closed as duplicate of #1189. If you have the same issue, please upvote or add a comment on #1189.
